### PR TITLE
Preserve additional location in non-source files when round tripping …

### LINKED ISSUE #46377: Additional document location added to `Diagnostic.AdditionalLocations` is not preserved
If an external location which points to a span in an additional document is added to `Diagnostic.AdditionalLocations`, this location is not preserved from an analyzer to code fix. This seems to be due to `IsInSource` check done here: https://github.com/dotnet/roslyn/blob/4fc1fa4bf205968b0dd556c502a79dbe80cc3519/src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs#L345-L350

_Originally posted by @mavasani in https://github.com/dotnet/roslyn-analyzers/pull/3914#discussion_r461221196_

### PR DESCRIPTION
**Pull request by mavasani**
…diagnostic

Fixes #46377

### CODE REVIEW

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs [hunk 1]
```diff
@@ -442,6 +436,36 @@ public static DiagnosticData Create(Diagnostic diagnostic, TextDocument document
...
+
+            using var _ = ArrayBuilder<DiagnosticDataLocation>.GetInstance(diagnostic.AdditionalLocations.Count, out var builder);
+            foreach (var location in diagnostic.AdditionalLocations)
+            {
+                if (location.IsInSource)
+                {
+                    if (CreateLocation(document.Project.GetDocument(location.SourceTree), location) is { } dataLocation)
+                    {
+                        builder.Add(dataLocation);
+                    }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs:454]:
```suggestion
                        builder.AddIfNotNull(CreateLocation(document.Project.GetDocument(location.SourceTree), location));
```

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs [hunk 2]
```diff
@@ -442,6 +436,36 @@ public static DiagnosticData Create(Diagnostic diagnostic, TextDocument document
...
+                }
+                else if (location.Kind == LocationKind.ExternalFile)
+                {
+                    if (document.Project.GetDocumentForExternalLocation(location) is { } textDocumentId &&
+                        CreateLocation(document.Project.GetTextDocument(textDocumentId), location) is { } dataLocation)
+                    {
+                        builder.Add(dataLocation);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs:461]:
```suggestion
                        builder.AddIfNotNull(CreateLocation(document.Project.GetTextDocument(document.Project.GetDocumentForExternalLocation(location)), location));
```

**Comment by mavasani** [comment]
Thanks @CyrusNajmabadi. Addressed feedback and also rebased the PR to target release/dev16.8

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs [hunk 3]
```diff
@@ -442,6 +436,32 @@ public static DiagnosticData Create(Diagnostic diagnostic, TextDocument document
...
+            using var _ = ArrayBuilder<DiagnosticDataLocation>.GetInstance(diagnostic.AdditionalLocations.Count, out var builder);
+            foreach (var location in diagnostic.AdditionalLocations)
+            {
+                if (location.IsInSource)
+                {
+                    builder.AddIfNotNull(CreateLocation(document.Project.GetDocument(location.SourceTree), location));
+                }
+                else if (location.Kind == LocationKind.ExternalFile)
+                {
+                    if (document.Project.GetDocumentForExternalLocation(location) is { } textDocumentId)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs:455]: i don't think this check is needed.  GetTextDocument allows null doc-id, right?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs (hunk 3 above)
Comment (mavasani) [reply src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs]: Thanks, done.

**Pull request merged**